stl/src/atomic.cpp

```diff
@@ -33,4 +39,281 @@ _CRTIMP2_PURE void __cdecl _Unlock_shared_ptr_spin_lock() { // release previousl
 #endif // _M_ARM
 }
 
+static std::once_flag _Atomic_spin_count_initialized;
+
+static long _Atomic_spin_count;
+
+static void _Atomic_spin_count_initialize() {
+    _Atomic_spin_count = (std::thread::hardware_concurrency() == 1 ? 0 : 10'000);

```

Comment:
Ditto this should not call_once even if you do want to cache it:

```
// given atomic<long> _Atomic_spin_count;

long _Local = _Atomic_spin_count.load(memory_order_relaxed);
if (_Local != 0) { return _Local; }
_Local = std::thread::hardware_concurrency() == 1 ? 0 : 10'000;
_Atomic_spin_count.store(_Local, memory_order_relaxed);
return _Local;
```

Reply:
> Given your other PR that makes this not a constant should it really be cached like this?

I agree that it could be changed at run-time by setting affinity.

But on the other hand, a kernel call on fast path is not welcome. (`GetNativeSystemInfo` and `GetProcessAffinityMask` are both kernel calls).

Need a way to make the decision on spinning cheap.

Reply:
That's fine, just wanted to make sure that decision was being made explicitly.

Reply:
The biggest boomer here is that I'm sure they are available without a kernel call.
WinDbg command "`dt nt!_PEB`" shows:
```
...
   +0x0b8 NumberOfProcessors : Uint4B
...
   +0x138 ActiveProcessAffinityMask : Uint8B
...
```
but I see no "legal" ways to reach them.

Possibly, getting HW concurrency count can be implemented via `__cpuidex` (though ignoring affinity)

Anyway, it is only for a fallback for pre-Win8 or low resources, and changing process affinity of runtime is a corner case. So it is performance of double corner case, don't think it worth much concern.

Reply:
Agreed.